File: Javascript/DSA/linked-list.js
"use strict"

let { Node } = require("./node");

function LinkedList() {
    let node = Node;

    let length = 0;
    let head = null;

    this.append = function(element){ 
        let node = new Node(element);

        if (head == null) {
            head = node;
        }
        else {
            let current = head;
            while (current.next) {
                current = current.next;
            }
            current.next = node;
        }
        length++;
     }

    this.insert = function(position, element) {
        if (position >= 0 && position <= length) {
            let node = new Node(element);
            current = head;
            previous = null;
            index = 0;

            if (position === 0) {
                node.next = current;
                head = node;
            }
            else {
                while (index++ < position) {
                    previous = current;
                    current = current.next;
                }
                node.next = current;
                previous.next = node;
            }
        }
    }
    this.removeAt = function(position){
        if (position > -1 && position < length) {
            let current = head;
            let previous = null;
            let index = 0;

            if (position === 0) {
                head = current.next;
            }
            else {
                while (index++ < position) {
                    previous = current;
                    current = current.next;
                }

                previous.next = current.next;
            }
            length--;
            return current.element;
        } else {
            return null;
        }
    }
    this.remove = function(element) {}
    this.indexOf  = function(element) {}
    this.isEmpty = function() {}
    this.size = function() { return length; }
    
    this.toString = function() { 
        let current = head;
        let string = '';
        while (current) {
            string = current.element + ' -> ' + string;
            current = current.next;
        }
        return string;
    }
    this.print = function() { console.log(this.toString())}   
}

let list = new LinkedList();
console.log(list.size());
list.append(99);
list.append(22);
list.print();

File: C#/DSA/LinkedList.cs
namespace DSA
{
    internal class Node<T>
    {
        public T Value { get; }
        public Node<T> Next { get; set; }
        public Node(T value)
        {
            Value = value;
        }
    }
    public class LinkedList<T>
    {
        private Node<T> head;
        private int count;

        public void Add(T value)
        {
            var node = new Node<T>(value);
            if (head == null)
            {
                head = node;
            }
            else
            {
                var current = head;
                while (current.Next != null)
                {
                    current = current.Next;
                }
                current.Next = node;
            }
            count++;
        }

        public bool Insert(int position, T value)
        {
            if (position >= 0 && position <= count)
            {
                var node = new Node<T>(value);
                var current = head;
                var index = 0;
                Node<T> previous = null;

                if (position == 0)
                {
                    node.Next = current;
                    head = node;
                }
                else
                {
                    while (index++ < position)
                    {
                        previous = current;
                        current = current.Next;
                    }
                    node.Next = current;
                    previous.Next = node;
                }
                count++;
                return true;
            }
            return false;
        }

        public bool RemoveAt(int position)
        {
            if (position > -1 && position < count)
            {
                var current = head;
                var index = 0;
                Node<T> previous = null;

                if (position == 0)
                {
                    head = current.Next;
                }
                else
                {
                    while (index++ < position)
                    {
                        previous = current;
                        current = current.Next;
                    }
                    previous.Next = current.Next;
                }
                count--;
                return true;
            }
            return false;
        }
        public bool Remove(T value)
        {
            var index = IndexOf(value);
            return RemoveAt(index);
        }
        public int IndexOf(T value)
        {
            var current = head;
            var index = 0;
            while (current != null)
            {
                if (current.Value.Equals(value))
                {
                    return index;
                }
                index++;
                current = current.Next;
            }
            return -1;
        }

        public bool IsEmpty() { return count == 0; }
        public int Size() { return count; }
        public T GetHead() { return head.Value; }

        public bool Contains(T value)
        {
            var node = head;
            while (node != null && !node.Value.Equals(value))
            {
                node = node.Next;
            }
            if (node == null)
            {
                return false;
            }
            return true;
        }

        public string Print()
        {
            string str = " ";
            if (head != null)
            {
                var current = head;

                while (current != null)
                {
                    str = str + current.Value.ToString() + " -> ";
                    current = current.Next;
                }
            }
            return str;
        }
    }
}
